File: 10-GRAPHS/Python/number-of-islands.py
from typing import List


# https://leetcode.com/problems/number-of-islands/
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if len(grid) == 0:
            return 0

        rows, cols = len(grid), len(grid[0])

        def dfs(r: int, c: int):
            if min(r, c) < 0 or r >= rows or c >= cols or grid[r][c] == '0':
                return

            grid[r][c] = '0'
            dfs(r - 1, c)
            dfs(r + 1, c)
            dfs(r, c - 1)
            dfs(r, c + 1)

        result = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '1':
                    result += 1
                    dfs(r, c)

        return result

File: 10-GRAPHS/PHP/number-of-islands.php
<?php

// https://leetcode.com/problems/number-of-islands/

class Solution
{
    private array $grid = [];
    private int $rows = 0;
    private int $cols = 0;

    /**
     * @param string[][] $grid
     * @return int
     */
    function numIslands(array $grid): int
    {
        if (count($grid) == 0) {
            return 0;
        }

        $this->grid = $grid;
        $this->rows = count($grid);
        $this->cols = count($grid[0]);

        $count = 0;
        for ($r = 0; $r < $this->rows; $r++) {
            for ($c = 0; $c < $this->cols; $c++) {
                if ($this->grid[$r][$c] == '1') {
                    $count++;
                    $this->dfs($r, $c);
                }
            }
        }
        return $count;
    }

    private function dfs(int $r, int $c): void
    {
        if ($this->isOutOfBound($r, $c) || $this->grid[$r][$c] == '0') {
            return;
        }

        $this->grid[$r][$c] = '0';
        $this->dfs($r-1, $c);
        $this->dfs($r+1, $c);
        $this->dfs($r, $c-1);
        $this->dfs($r, $c+1);
    }

    private function isOutOfBound(int $r, int $c): bool
    {
        return min($r, $c) < 0 || $r >= $this->rows || $c >= $this->cols;
    }

}